tensorstore/kvstore/s3/s3_request_builder.cc

```diff
@@ -0,0 +1,285 @@
+
+  auto & request_ = builder_.request_;
+
+  auto url = std::string_view(request_.url());
+  std::vector<std::pair<std::string, std::string>> queries;
+
+  if(auto query_pos = url.find('?'); query_pos != std::string::npos && query_pos + 1 != std::string::npos) {
+    auto query = url.substr(query_pos + 1);
+    std::vector<std::string_view> query_bits = absl::StrSplit(query, '&');
+
+    for(auto query_bit: query_bits) {
+      auto pos = query_bit.find("=");
+      auto key = query_bit.substr(0, pos);
+      auto value = query_bit.substr(pos + 1, std::string_view::npos);
+      queries.push_back({std::string(key), std::string(value)});
+    }
+  }
+
+  auto headers = request_.headers();
+  std::sort(std::begin(headers), std::end(headers));
+  std::sort(std::begin(queries), std::end(queries));
+  auto canonical_request = CanonicalRequest(request_.url(), request_.method(),
+                                            payload_hash, headers,
+                                            queries);
+  auto signing_string = SigningString(canonical_request, aws_region, time);
+  auto signature = Signature(aws_secret_access_key, aws_region, signing_string, time);
+  auto auth_header = AuthorizationHeader(aws_access_key, aws_region, signature, headers, time);
+  request_.headers_.emplace_back(std::move(auth_header));
+  return std::move(request_);
+}
+
+/// https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+std::string S3RequestBuilder::SigningString(
+  std::string_view canonical_request,
+  std::string_view aws_region,
+  const absl::Time & time)
+{
+  absl::TimeZone utc = absl::UTCTimeZone();
+  SHA256Digester sha256;
+  sha256.Write(canonical_request);
+  const auto digest = sha256.Digest();
+  auto digest_sv = std::string_view(
+    reinterpret_cast<const char *>(digest.begin()),
+    digest.size());
+
+  return absl::StrFormat(
+    "AWS4-HMAC-SHA256\n"
+    "%s\n"
+    "%s/%s/s3/aws4_request\n"
+    "%s",
+      absl::FormatTime("%Y%m%dT%H%M%SZ", time, utc),
+      absl::FormatTime("%Y%m%d", time, utc), aws_region,
+      absl::BytesToHexString(digest_sv));
+}
+
+std::string S3RequestBuilder::Signature(
+  std::string_view aws_secret_access_key,
+  std::string_view aws_region,
+  std::string_view signing_string,
+  const absl::Time & time)
+{
+  absl::TimeZone utc = absl::UTCTimeZone();
+  unsigned char date_key[kHmacSize];
+  unsigned char date_region_key[kHmacSize];
+  unsigned char date_region_service_key[kHmacSize];
+  unsigned char signing_key[kHmacSize];
+  unsigned char final_key[kHmacSize];
+
+  ComputeHmac(absl::StrCat("AWS4",aws_secret_access_key),
+              absl::FormatTime("%Y%m%d", time, utc), date_key);
+  ComputeHmac(date_key, aws_region, date_region_key);
+  ComputeHmac(date_region_key, "s3", date_region_service_key);
+  ComputeHmac(date_region_service_key, "aws4_request", signing_key);
+  ComputeHmac(signing_key, signing_string, final_key);
+
+  std::string result(2 * kHmacSize, '0');
```

Comment:
So this is probably a better option:

#include "absl/strings/ascii.h"

std::string result = absl::BytesToHexString(final_key)
absl::AsciiStrToUpper(result);

Reply:
done (although didn't need conversion to uppercase)